File: frontend/src/api/audit.ts
/**
 * Audit Chain API Client
 * ======================
 *
 * TypeScript API client functions for the E2I Audit Chain endpoints.
 * Uses the shared apiClient for consistent error handling and interceptors.
 *
 * The audit chain provides cryptographic proof of agent execution
 * sequence and data integrity using hash chains.
 *
 * Endpoints:
 * - Workflow entry retrieval
 * - Chain integrity verification
 * - Workflow summary aggregation
 * - Recent workflow listing
 *
 * @module api/audit
 */

import { get } from '@/lib/api-client';
import type {
  AuditEntryResponse,
  ChainVerificationResponse,
  ListRecentWorkflowsParams,
  ListWorkflowEntriesParams,
  RecentWorkflowResponse,
  WorkflowSummaryResponse,
} from '@/types/audit';

// =============================================================================
// AUDIT CHAIN API ENDPOINTS
// =============================================================================

const AUDIT_BASE = '/audit';

// =============================================================================
// WORKFLOW ENTRY ENDPOINTS
// =============================================================================

/**
 * Get all audit entries for a workflow.
 *
 * Retrieves audit chain entries ordered by sequence number.
 * Each entry contains agent action, hash chain data, and metrics.
 *
 * @param workflowId - Workflow UUID
 * @param params - Optional pagination parameters
 * @returns List of audit entries in sequence order
 *
 * @example
 * ```typescript
 * const entries = await getWorkflowEntries('abc123-uuid');
 * entries.forEach(entry => {
 *   console.log(`[${entry.sequence_number}] ${entry.agent_name}: ${entry.action_type}`);
 * });
 * ```
 */
export async function getWorkflowEntries(
  workflowId: string,
  params?: ListWorkflowEntriesParams
): Promise<AuditEntryResponse[]> {
  return get<AuditEntryResponse[]>(
    `${AUDIT_BASE}/workflow/${encodeURIComponent(workflowId)}`,
    params as Record<string, unknown> | undefined
  );
}

/**
 * Verify the integrity of a workflow's audit chain.
 *
 * Performs cryptographic verification by validating hash chains
 * between consecutive entries.
 *
 * @param workflowId - Workflow UUID to verify
 * @returns Verification result with validity status
 *
 * @example
 * ```typescript
 * const result = await verifyWorkflowChain('abc123-uuid');
 * if (result.is_valid) {
 *   console.log(`Chain verified: ${result.entries_checked} entries`);
 * } else {
 *   console.error(`Chain broken at entry: ${result.first_invalid_entry}`);
 *   console.error(`Error: ${result.error_message}`);
 * }
 * ```
 */
export async function verifyWorkflowChain(
  workflowId: string
): Promise<ChainVerificationResponse> {
  return get<ChainVerificationResponse>(
    `${AUDIT_BASE}/workflow/${encodeURIComponent(workflowId)}/verify`
  );
}

/**
 * Get a summary of a workflow's audit chain.
 *
 * Returns aggregated metrics including agents involved, tiers,
 * total duration, and validation statistics.
 *
 * @param workflowId - Workflow UUID
 * @returns Workflow summary with aggregated metrics
 *
 * @example
 * ```typescript
 * const summary = await getWorkflowSummary('abc123-uuid');
 * console.log(`Workflow: ${summary.workflow_id}`);
 * console.log(`Entries: ${summary.total_entries}`);
 * console.log(`Agents: ${summary.agents_involved.join(', ')}`);
 * console.log(`Duration: ${summary.total_duration_ms}ms`);
 * console.log(`Chain valid: ${summary.chain_verified}`);
 * ```
 */
export async function getWorkflowSummary(
  workflowId: string
): Promise<WorkflowSummaryResponse> {
  return get<WorkflowSummaryResponse>(
    `${AUDIT_BASE}/workflow/${encodeURIComponent(workflowId)}/summary`
  );
}

// =============================================================================
// RECENT WORKFLOWS ENDPOINTS
// =============================================================================

/**
 * Get a list of recent audit workflows.
 *
 * Returns workflows with basic info, ordered by start time descending.
 * Can filter by brand or agent name.
 *
 * @param params - Optional filter and pagination parameters
 * @returns List of recent workflows
 *
 * @example
 * ```typescript
 * // Get all recent workflows
 * const recent = await getRecentWorkflows({ limit: 10 });
 * recent.forEach(w => {
 *   console.log(`${w.workflow_id}: ${w.first_agent} -> ${w.last_agent}`);
 * });
 *
 * // Filter by brand
 * const brandWorkflows = await getRecentWorkflows({
 *   brand: 'Remibrutinib',
 *   limit: 20,
 * });
 * ```
 */
export async function getRecentWorkflows(
  params?: ListRecentWorkflowsParams
): Promise<RecentWorkflowResponse[]> {
  return get<RecentWorkflowResponse[]>(`${AUDIT_BASE}/recent`, params as Record<string, unknown> | undefined);
}

// =============================================================================
// CONVENIENCE FUNCTIONS
// =============================================================================

/**
 * Get workflow details including entries, summary, and verification.
 *
 * Fetches all workflow data in parallel for comprehensive view.
 *
 * @param workflowId - Workflow UUID
 * @returns Complete workflow data
 *
 * @example
 * ```typescript
 * const workflow = await getWorkflowDetails('abc123-uuid');
 * console.log(`Summary: ${workflow.summary.total_entries} entries`);
 * console.log(`Verified: ${workflow.verification.is_valid}`);
 * workflow.entries.forEach(e => console.log(e.agent_name));
 * ```
 */
export async function getWorkflowDetails(workflowId: string): Promise<{
  entries: AuditEntryResponse[];
  summary: WorkflowSummaryResponse;
  verification: ChainVerificationResponse;
}> {
  const [entries, summary, verification] = await Promise.all([
    getWorkflowEntries(workflowId),
    getWorkflowSummary(workflowId),
    verifyWorkflowChain(workflowId),
  ]);

  return { entries, summary, verification };
}

/**
 * Get workflow by first agent.
 *
 * Filters recent workflows to find those starting with a specific agent.
 *
 * @param agentName - Agent name to filter by
 * @param limit - Maximum workflows to return
 * @returns Workflows starting with the specified agent
 *
 * @example
 * ```typescript
 * const orchestratorWorkflows = await getWorkflowsByFirstAgent('orchestrator', 10);
 * console.log(`Found ${orchestratorWorkflows.length} workflows`);
 * ```
 */
export async function getWorkflowsByFirstAgent(
  agentName: string,
  limit: number = 20
): Promise<RecentWorkflowResponse[]> {
  const workflows = await getRecentWorkflows({ agent_name: agentName, limit });
  return workflows.filter((w) => w.first_agent === agentName);
}

/**
 * Get workflow by brand.
 *
 * Filters recent workflows by pharmaceutical brand.
 *
 * @param brand - Brand name to filter by
 * @param limit - Maximum workflows to return
 * @returns Workflows for the specified brand
 *
 * @example
 * ```typescript
 * const remibWorkflows = await getWorkflowsByBrand('Remibrutinib', 10);
 * console.log(`Found ${remibWorkflows.length} Remibrutinib workflows`);
 * ```
 */
export async function getWorkflowsByBrand(
  brand: string,
  limit: number = 20
): Promise<RecentWorkflowResponse[]> {
  return getRecentWorkflows({ brand, limit });
}

/**
 * Check if a workflow chain is valid.
 *
 * Simple boolean check for chain integrity.
 *
 * @param workflowId - Workflow UUID
 * @returns True if chain is valid, false otherwise
 *
 * @example
 * ```typescript
 * const isValid = await isWorkflowChainValid('abc123-uuid');
 * if (!isValid) {
 *   alert('Workflow audit chain has been tampered with!');
 * }
 * ```
 */
export async function isWorkflowChainValid(workflowId: string): Promise<boolean> {
  try {
    const result = await verifyWorkflowChain(workflowId);
    return result.is_valid;
  } catch {
    return false;
  }
}

/**
 * Get agent execution path for a workflow.
 *
 * Returns the sequence of agents that executed in the workflow.
 *
 * @param workflowId - Workflow UUID
 * @returns Array of agent names in execution order
 *
 * @example
 * ```typescript
 * const path = await getAgentExecutionPath('abc123-uuid');
 * console.log(`Execution path: ${path.join(' -> ')}`);
 * // "orchestrator -> causal_impact -> explainer"
 * ```
 */
export async function getAgentExecutionPath(workflowId: string): Promise<string[]> {
  const entries = await getWorkflowEntries(workflowId);
  return entries.map((e) => e.agent_name);
}

/**
 * Get tier distribution for a workflow.
 *
 * Returns count of entries by agent tier.
 *
 * @param workflowId - Workflow UUID
 * @returns Map of tier number to entry count
 *
 * @example
 * ```typescript
 * const distribution = await getTierDistribution('abc123-uuid');
 * Object.entries(distribution).forEach(([tier, count]) => {
 *   console.log(`Tier ${tier}: ${count} entries`);
 * });
 * ```
 */
export async function getTierDistribution(
  workflowId: string
): Promise<Record<number, number>> {
  const entries = await getWorkflowEntries(workflowId);
  const distribution: Record<number, number> = {};

  for (const entry of entries) {
    distribution[entry.agent_tier] = (distribution[entry.agent_tier] || 0) + 1;
  }

  return distribution;
}

/**
 * Get total duration for a workflow.
 *
 * Sums up all entry durations.
 *
 * @param workflowId - Workflow UUID
 * @returns Total duration in milliseconds
 *
 * @example
 * ```typescript
 * const duration = await getWorkflowDuration('abc123-uuid');
 * console.log(`Total duration: ${duration}ms`);
 * ```
 */
export async function getWorkflowDuration(workflowId: string): Promise<number> {
  const summary = await getWorkflowSummary(workflowId);
  return summary.total_duration_ms;
}

/**
 * Get workflow time range.
 *
 * Returns the start and end timestamps for a workflow.
 *
 * @param workflowId - Workflow UUID
 * @returns Start and end timestamps
 *
 * @example
 * ```typescript
 * const range = await getWorkflowTimeRange('abc123-uuid');
 * console.log(`Started: ${range.start}`);
 * console.log(`Ended: ${range.end}`);
 * ```
 */
export async function getWorkflowTimeRange(workflowId: string): Promise<{
  start: string | null;
  end: string | null;
}> {
  const summary = await getWorkflowSummary(workflowId);
  return {
    start: summary.first_entry_at || null,
    end: summary.last_entry_at || null,
  };
}

/**
 * Get entries with failed validation.
 *
 * Filters workflow entries to only those that failed validation.
 *
 * @param workflowId - Workflow UUID
 * @returns Entries that failed validation
 *
 * @example
 * ```typescript
 * const failed = await getFailedValidationEntries('abc123-uuid');
 * if (failed.length > 0) {
 *   console.warn(`${failed.length} entries failed validation`);
 *   failed.forEach(e => console.warn(`  ${e.agent_name}: ${e.action_type}`));
 * }
 * ```
 */
export async function getFailedValidationEntries(
  workflowId: string
): Promise<AuditEntryResponse[]> {
  const entries = await getWorkflowEntries(workflowId);
  return entries.filter((e) => e.validation_passed === false);
}

/**
 * Get entries with low confidence.
 *
 * Filters workflow entries to those with confidence below threshold.
 *
 * @param workflowId - Workflow UUID
 * @param threshold - Confidence threshold (0-1, default: 0.7)
 * @returns Entries with confidence below threshold
 *
 * @example
 * ```typescript
 * const lowConfidence = await getLowConfidenceEntries('abc123-uuid', 0.8);
 * lowConfidence.forEach(e => {
 *   console.log(`${e.agent_name}: ${e.confidence_score}`);
 * });
 * ```
 */
export async function getLowConfidenceEntries(
  workflowId: string,
  threshold: number = 0.7
): Promise<AuditEntryResponse[]> {
  const entries = await getWorkflowEntries(workflowId);
  return entries.filter(
    (e) => e.confidence_score !== undefined && e.confidence_score < threshold
  );
}

/**
 * Format duration for display.
 *
 * Converts milliseconds to human-readable format.
 *
 * @param durationMs - Duration in milliseconds
 * @returns Formatted duration string
 *
 * @example
 * ```typescript
 * console.log(formatDuration(1500)); // "1.5s"
 * console.log(formatDuration(150)); // "150ms"
 * console.log(formatDuration(65000)); // "1m 5s"
 * ```
 */
export function formatDuration(durationMs: number): string {
  if (durationMs < 1000) {
    return `${durationMs}ms`;
  }
  if (durationMs < 60000) {
    return `${(durationMs / 1000).toFixed(1)}s`;
  }
  const minutes = Math.floor(durationMs / 60000);
  const seconds = Math.round((durationMs % 60000) / 1000);
  return `${minutes}m ${seconds}s`;
}

/**
 * Get entry hash chain status.
 *
 * Checks if each entry's hash chain is properly linked.
 *
 * @param entries - Array of audit entries
 * @returns Array of boolean indicating if each entry is properly linked
 *
 * @example
 * ```typescript
 * const entries = await getWorkflowEntries('abc123-uuid');
 * const chainStatus = getHashChainStatus(entries);
 * const brokenLinks = chainStatus.filter(s => !s.valid);
 * if (brokenLinks.length > 0) {
 *   console.error('Chain has broken links!');
 * }
 * ```
 */
export function getHashChainStatus(entries: AuditEntryResponse[]): Array<{
  entry_id: string;
  sequence_number: number;
  valid: boolean;
}> {
  return entries.map((entry, index) => {
    const valid =
      index === 0
        ? entry.previous_entry_id === undefined || entry.previous_entry_id === null
        : entry.previous_entry_id === entries[index - 1].entry_id;

    return {
      entry_id: entry.entry_id,
      sequence_number: entry.sequence_number,
      valid,
    };
  });
}

File: src/api/routes/audit.py
"""
E2I Audit Chain API
====================

FastAPI endpoints for audit chain verification and inspection.

Endpoints:
- GET /audit/workflow/{workflow_id}: Get all entries for a workflow
- GET /audit/workflow/{workflow_id}/verify: Verify chain integrity
- GET /audit/workflow/{workflow_id}/summary: Get workflow summary
- GET /audit/recent: Get recent audit workflows

Integration Points:
- AuditChainService for chain operations
- Supabase for persistence
- All 18 agents emit audit entries

Author: E2I Causal Analytics Team
Version: 4.1.0
"""

import logging
from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, ConfigDict

from src.api.dependencies.auth import require_auth
from src.api.schemas.errors import ErrorResponse, ValidationErrorResponse
from src.utils.audit_chain import (
    AuditChainService,
)
from src.utils.type_helpers import parse_supabase_row, parse_supabase_rows

logger = logging.getLogger(__name__)

router = APIRouter(
    prefix="/audit",
    tags=["Audit Chain"],
    responses={
        401: {"model": ErrorResponse, "description": "Authentication required"},
        422: {"model": ValidationErrorResponse, "description": "Validation error"},
        500: {"model": ErrorResponse, "description": "Internal server error"},
    },
)


# =============================================================================
# RESPONSE MODELS
# =============================================================================


class AuditEntryResponse(BaseModel):
    """Response model for a single audit chain entry."""

    entry_id: UUID
    workflow_id: UUID
    sequence_number: int
    agent_name: str
    agent_tier: int
    action_type: str
    created_at: datetime
    duration_ms: Optional[int] = None

    # Validation & confidence
    validation_passed: Optional[bool] = None
    confidence_score: Optional[float] = None
    refutation_results: Optional[Dict[str, Any]] = None

    # Hash chain fields
    previous_entry_id: Optional[UUID] = None
    previous_hash: Optional[str] = None
    entry_hash: str

    # Context
    user_id: Optional[str] = None
    session_id: Optional[UUID] = None
    brand: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class ChainVerificationResponse(BaseModel):
    """Response model for chain verification result."""

    workflow_id: UUID
    is_valid: bool
    entries_checked: int
    first_invalid_entry: Optional[UUID] = None
    error_message: Optional[str] = None
    verified_at: datetime

    model_config = ConfigDict(from_attributes=True)


class WorkflowSummaryResponse(BaseModel):
    """Response model for workflow summary."""

    workflow_id: UUID
    total_entries: int
    first_entry_at: Optional[datetime] = None
    last_entry_at: Optional[datetime] = None
    agents_involved: List[str]
    tiers_involved: List[int]
    chain_verified: bool
    brand: Optional[str] = None

    # Aggregated metrics
    total_duration_ms: int = 0
    avg_confidence_score: Optional[float] = None
    validation_passed_count: int = 0
    validation_failed_count: int = 0


class RecentWorkflowResponse(BaseModel):
    """Response model for recent workflow listing."""

    workflow_id: UUID
    started_at: datetime
    entry_count: int
    first_agent: str
    last_agent: str
    brand: Optional[str] = None


# =============================================================================
# DEPENDENCY - Get Audit Service
# =============================================================================


def get_audit_service() -> Optional[AuditChainService]:
    """
    Get the global AuditChainService instance.

    Returns None if service is not initialized (graceful degradation).
    """
    from src.agents.base.audit_chain_mixin import get_audit_chain_service

    return get_audit_chain_service()


# =============================================================================
# ENDPOINTS
# =============================================================================


@router.get(
    "/workflow/{workflow_id}",
    response_model=List[AuditEntryResponse],
    summary="Get workflow audit entries",
    operation_id="get_workflow_entries",
    description="Retrieve all audit chain entries for a specific workflow, ordered by sequence number.",
)
async def get_workflow_entries(
    workflow_id: UUID,
    limit: int = Query(default=100, ge=1, le=1000, description="Maximum entries to return"),
    offset: int = Query(default=0, ge=0, description="Number of entries to skip"),
    user: Dict[str, Any] = Depends(require_auth),
) -> List[AuditEntryResponse]:
    """Get all audit entries for a workflow."""
    service = get_audit_service()
    if service is None:
        raise HTTPException(
            status_code=503,
            detail="Audit chain service unavailable. Check service configuration.",
        )

    try:
        # Query entries from database
        result = (
            service.db.table("audit_chain_entries")
            .select("*")
            .eq("workflow_id", str(workflow_id))
            .order("sequence_number")
            .range(offset, offset + limit - 1)
            .execute()
        )

        if not result.data:
            return []

        # Convert to response models
        entries = []
        for row in parse_supabase_rows(result.data):
            entries.append(_row_to_response(row))

        return entries

    except Exception as e:
        logger.error(f"Failed to get workflow entries: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to retrieve entries: {str(e)}")


@router.get(
    "/workflow/{workflow_id}/verify",
    response_model=ChainVerificationResponse,
    summary="Verify chain integrity",
    operation_id="verify_workflow_chain",
    description="Verify the cryptographic integrity of a workflow's audit chain.",
)
async def verify_workflow_chain(
    workflow_id: UUID,
    user: Dict[str, Any] = Depends(require_auth),
) -> ChainVerificationResponse:
    """Verify the integrity of a workflow's audit chain."""
    service = get_audit_service()
    if service is None:
        raise HTTPException(
            status_code=503,
            detail="Audit chain service unavailable. Check service configuration.",
        )

    try:
        verification = service.verify_workflow(workflow_id)

        return ChainVerificationResponse(
            workflow_id=workflow_id,
            is_valid=verification.is_valid,
            entries_checked=verification.entries_checked,
            first_invalid_entry=verification.first_invalid_entry,
            error_message=verification.error_message,
            verified_at=verification.verified_at,
        )

    except Exception as e:
        logger.error(f"Failed to verify workflow: {e}")
        raise HTTPException(status_code=500, detail=f"Verification failed: {str(e)}")


@router.get(
    "/workflow/{workflow_id}/summary",
    response_model=WorkflowSummaryResponse,
    summary="Get workflow summary",
    operation_id="get_workflow_summary",
    description="Get a summary of a workflow including agents involved and aggregated metrics.",
)
async def get_workflow_summary(
    workflow_id: UUID,
    user: Dict[str, Any] = Depends(require_auth),
) -> WorkflowSummaryResponse:
    """Get a summary of a workflow's audit chain."""
    service = get_audit_service()
    if service is None:
        raise HTTPException(
            status_code=503,
            detail="Audit chain service unavailable. Check service configuration.",
        )

    try:
        # Get all entries for the workflow
        result = (
            service.db.table("audit_chain_entries")
            .select("*")
            .eq("workflow_id", str(workflow_id))
            .order("sequence_number")
            .execute()
        )

        if not result.data:
            raise HTTPException(status_code=404, detail=f"Workflow {workflow_id} not found")

        entries = parse_supabase_rows(result.data)

        # Aggregate data
        agents = list({e["agent_name"] for e in entries})
        tiers = list({e["agent_tier"] for e in entries})

        total_duration = sum(e.get("duration_ms") or 0 for e in entries)

        confidence_scores = [
            e["confidence_score"] for e in entries if e.get("confidence_score") is not None
        ]
        avg_confidence = (
            sum(confidence_scores) / len(confidence_scores) if confidence_scores else None
        )

        validation_passed = sum(1 for e in entries if e.get("validation_passed") is True)
        validation_failed = sum(1 for e in entries if e.get("validation_passed") is False)

        # Check chain validity
        try:
            verification = service.verify_workflow(workflow_id)
            chain_verified = verification.is_valid
        except Exception:
            chain_verified = False

        # Get brand from first entry
        brand = entries[0].get("brand") if entries else None

        return WorkflowSummaryResponse(
            workflow_id=workflow_id,
            total_entries=len(entries),
            first_entry_at=datetime.fromisoformat(entries[0]["created_at"].replace("Z", "+00:00")),
            last_entry_at=datetime.fromisoformat(entries[-1]["created_at"].replace("Z", "+00:00")),
            agents_involved=sorted(agents),
            tiers_involved=sorted(tiers),
            chain_verified=chain_verified,
            brand=brand,
            total_duration_ms=total_duration,
            avg_confidence_score=avg_confidence,
            validation_passed_count=validation_passed,
            validation_failed_count=validation_failed,
        )

    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Failed to get workflow summary: {e}")
        raise HTTPException(status_code=500, detail=f"Failed to get summary: {str(e)}")


@router.get(
    "/recent",
    response_model=List[RecentWorkflowResponse],
    summary="Get recent workflows",
    operation_id="get_recent_workflows",
    description="Get a list of recent audit workflows with basic info.",
)
async def get_recent_workflows(
    limit: int = Query(default=20, ge=1, le=100, description="Maximum workflows to return"),
    brand: Optional[str] = Query(default=None, description="Filter by brand"),
    agent_name: Optional[str] = Query(default=None, description="Filter by agent name"),
    user: Dict[str, Any] = Depends(require_auth),
) -> List[RecentWorkflowResponse]:
    """Get recent audit workflows."""
    service = get_audit_service()
    if service is None:
        raise HTTPException(
            status_code=503,
            detail="Audit chain service unavailable. Check service configuration.",
        )

    try:
        # Use SQL view for efficient aggregation
        query = service.db.rpc(
            "get_recent_audit_workflows",
            {"p_limit": limit, "p_brand": brand, "p_agent_name": agent_name},
        )
        result = query.execute()

        if not result.data:
            # Fallback: direct query if RPC not available
            return await _get_recent_workflows_fallback(service, limit, brand, agent_name)

        return [
            RecentWorkflowResponse(
                workflow_id=UUID(row["workflow_id"]),
                started_at=datetime.fromisoformat(row["started_at"].replace("Z", "+00:00")),
                entry_count=row["entry_count"],
                first_agent=row["first_agent"],
                last_agent=row["last_agent"],
                brand=row.get("brand"),
            )
            for row in parse_supabase_rows(result.data)
        ]

    except Exception as e:
        logger.warning(f"RPC failed, using fallback: {e}")
        return await _get_recent_workflows_fallback(service, limit, brand, agent_name)


# =============================================================================
# HELPER FUNCTIONS
# =============================================================================


def _row_to_response(row: Dict[str, Any]) -> AuditEntryResponse:
    """Convert a database row to AuditEntryResponse."""
    return AuditEntryResponse(
        entry_id=UUID(row["entry_id"]),
        workflow_id=UUID(row["workflow_id"]),
        sequence_number=row["sequence_number"],
        agent_name=row["agent_name"],
        agent_tier=row["agent_tier"],
        action_type=row["action_type"],
        created_at=datetime.fromisoformat(row["created_at"].replace("Z", "+00:00")),
        duration_ms=row.get("duration_ms"),
        validation_passed=row.get("validation_passed"),
        confidence_score=row.get("confidence_score"),
        refutation_results=row.get("refutation_results"),
        previous_entry_id=(
            UUID(row["previous_entry_id"]) if row.get("previous_entry_id") else None
        ),
        previous_hash=row.get("previous_hash"),
        entry_hash=row["entry_hash"],
        user_id=row.get("user_id"),
        session_id=UUID(row["session_id"]) if row.get("session_id") else None,
        brand=row.get("brand"),
    )


async def _get_recent_workflows_fallback(
    service: AuditChainService,
    limit: int,
    brand: Optional[str],
    agent_name: Optional[str],
) -> List[RecentWorkflowResponse]:
    """Fallback method to get recent workflows without RPC."""
    try:
        # Get distinct workflow_ids with first entry
        query = service.db.table("audit_chain_entries").select(
            "workflow_id, created_at, agent_name, brand"
        )

        if brand:
            query = query.eq("brand", brand)
        if agent_name:
            query = query.eq("agent_name", agent_name)

        query = query.eq("sequence_number", 1).order("created_at", desc=True).limit(limit)

        result = query.execute()

        if not result.data:
            return []

        workflows = []
        for row in parse_supabase_rows(result.data):
            workflow_id = UUID(row["workflow_id"])

            # Get entry count and last agent
            count_result = (
                service.db.table("audit_chain_entries")
                .select("agent_name, sequence_number")
                .eq("workflow_id", str(workflow_id))
                .order("sequence_number", desc=True)
                .limit(1)
                .execute()
            )

            entry_count = 1
            last_agent = row["agent_name"]
            if count_result.data:
                count_row = parse_supabase_row(count_result.data[0])
                entry_count = count_row["sequence_number"]
                last_agent = count_row["agent_name"]

            workflows.append(
                RecentWorkflowResponse(
                    workflow_id=workflow_id,
                    started_at=datetime.fromisoformat(row["created_at"].replace("Z", "+00:00")),
                    entry_count=entry_count,
                    first_agent=row["agent_name"],
                    last_agent=last_agent,
                    brand=row.get("brand"),
                )
            )

        return workflows

    except Exception as e:
        logger.error(f"Fallback query failed: {e}")
        return []
